src/realm/sync/subscriptions.cpp

```diff
@@ -305,6 +306,64 @@ SubscriptionSet SubscriptionSet::make_mutable_copy() const
     return new_set_obj;
 }
 
+util::Future<void> SubscriptionSet::get_state_change_notification(State notify_when) const
+{
+    // If we've already reached the desired state, or if the subscription is in an error state,
+    // we can return a ready future immediately.
+    if (state() >= notify_when) {
+        return util::Future<void>::make_ready();
+    }
+    else if (state() == State::Error) {
+        return util::Future<void>::make_ready(Status{ErrorCodes::RuntimeError, error_str()});
+    }
+
+    std::lock_guard<std::mutex> lk(m_mgr->m_pending_notifications_mutex);
+
+    // If we've already been superceded by another version getting completed, then we should skip registering
+    // a notification because it may never fire.
+    if (m_mgr->m_min_outstanding_version > version()) {
+        return util::Future<void>::make_ready();
+    }
+
+    // Otherwise, make a promise/future pair and add it to the list of pending notifications.
+    auto [promise, future] = util::make_promise_future<void>();
+    m_mgr->m_pending_notifications.emplace_back(version(), std::move(promise), notify_when);
+    return std::move(future);
+}
+
+void SubscriptionSet::process_notifications()
+{
+    auto new_state = state();
+    auto my_version = version();
+
+    std::list<SubscriptionStore::NotificationRequest> to_finish;
+    std::unique_lock<std::mutex> lk(m_mgr->m_pending_notifications_mutex);
+    for (auto it = m_mgr->m_pending_notifications.begin(); it != m_mgr->m_pending_notifications.end();) {
+        if ((it->version == my_version && (new_state == State::Error || new_state >= it->notify_when)) ||
+            (new_state == State::Complete && it->version < my_version)) {
```

Comment:
So the intention here is - if this is a notification request for my version and either my state is an error (we always resolve errors as errors) or my state has advanced beyond the requested state then we should handle the notification request, and separately if I have been marked complete I should clean up notification requests for previous versions. Basically the server is allowed to skip versions, so it's possible to go from version 1 to version 3 and mark version 3 as complete without ever getting any notifications for version 2. This is supposed to take care of that.

Does that make more sense?

Reply:
Thanks for the explanation and sorry for being vague. To clarify, let's say we have a notification request for "complete" on version 2 but version 2 is skipped, and the server sets an error for version 3. In that case, this conditional will not resolve the notification on version 2; is that correct? I think I would expect the notification for version 2 to resolve with an error which I think my code suggestion will do.

Reply:
I think if I were a user and asked for a notification on version 2, i would be pretty annoyed if I got an unrelated error from version 3 and then also an error for version 3's notification.  My goal here was to try to make sure that we deliver error messages, which will likely be something like "cannot convert your query to MQL - please change this query on this table", to exactly one notification request, and to treat superceding a version as a special kind of success rather than an error. So maybe what we need here is a tri-state result for the notifications. Either there's an error or it will resolve as either completed or as superceded.

Reply:
Oh ok that makes sense, I didn't know that the server could generate an error for a later version while still populating an earlier version. I thought that any later versions would entirely supersede the previous ones. I think having a superceded state would clarify things in this code, but I'm not as sure it would make sense to expose it to the user. If you'd like to add a tri-state later on that's fine with me, but for now it looks like this is working as intended 👍